File: apps/api/app/schemas/contracts.py
from typing import Any
from pydantic import BaseModel

from app.schemas.common import DecisionBand, OutputMode, ZoningDecision


class ReliabilityIndicator(BaseModel):
    module_name: str
    overall_confidence_score: float
    freshness_score: float
    source_quality_score: float
    coverage_score: float
    model_stability_score: float
    missing_data_warning: bool
    observed_ratio: float
    estimated_ratio: float
    manual_review_recommended: bool
    explanation_note: str


class CategoryComparisonRow(BaseModel):
    category_code: str
    band: DecisionBand
    suitability_score: float
    saturation_score: float
    opportunity_score: float
    risk_level: str
    preferred_floor_type: str | None = None
    preferred_unit_size: str | None = None
    reason_summary: str
    confidence: float
    suggested_alternative: str | None = None


class CategoryComparisonResponse(BaseModel):
    project_id: str
    scenario_id: str | None = None
    mode: OutputMode
    rows: list[CategoryComparisonRow]
    pagination: dict[str, int]
    reliability_indicator: ReliabilityIndicator


class UnsuitableCategoryCard(BaseModel):
    category_code: str
    rejection_level: str
    poor_fit_reason: str
    saturation_reason: str | None = None
    demographic_mismatch_reason: str | None = None
    location_unit_mismatch_reason: str | None = None
    zoning_caution: str | None = None
    pricing_burden_warning: str | None = None
    alternatives: list[str] = []
    operator_explanation: str
    customer_explanation: str
    confidence: float


class ValidationAssessment(BaseModel):
    assessment_id: str
    tenant_validation_score: float
    decision_band: DecisionBand
    risk_level: str
    suitability_score: float
    saturation_score: float
    opportunity_score: float


class ValidationResultResponse(BaseModel):
    assessment: ValidationAssessment
    unsuitable_category_cards: list[UnsuitableCategoryCard]
    alternatives: list[str]
    reliability_indicator: ReliabilityIndicator
    manual_adjustment_effects: list[dict[str, Any]]
    zoning_summary: dict[str, Any]


class PricingRecommendation(BaseModel):
    unit_id: str
    recommended_deposit_krw: float
    recommended_monthly_rent_krw: float
    lower_bound_krw: float | None = None
    upper_bound_krw: float | None = None
    pricing_confidence_score: float
    assumptions: dict[str, Any]


class ManualAdjustmentCreateRequest(BaseModel):
    module_name: str
    field_name: str
    target_entity_type: str
    target_entity_id: str
    after_value: Any
    reason_code: str
    reason_note: str | None = None


class ManualAdjustmentCreateResponse(BaseModel):
    adjustment_id: str
    approval_required: bool
    recompute_job_id: str


class ZoningScreeningResult(BaseModel):
    rule_id: str
    decision_status: ZoningDecision
    operator_message: str
    customer_message: str
    legal_basis_text: str | None = None
    legal_reference_code: str | None = None
    review_required: bool
    review_owner_user_id: str | None = None


class ZoningScreeningResponse(BaseModel):
    run_id: str
    status: str
    results: list[ZoningScreeningResult]


class CRMLeadSummary(BaseModel):
    lead_id: str
    lead_type: str
    lead_stage: str
    lead_score: float
    validation_assessment_id: str | None = None
    report_run_id: str | None = None
    next_followup_at: str | None = None


class ReportSummary(BaseModel):
    report_run_id: str
    mode: OutputMode
    status: str
    section_status: list[dict[str, Any]]
    evidence_coverage_ratio: float
    html_uri: str | None = None
    pdf_uri: str | None = None


class AnalysisJobStatus(BaseModel):
    job_id: str
    module: str
    status: str
    percent_progress: int
    partial_available: bool
    partial_payload_uri: str | None = None
    retry_token: str | None = None
    error: dict[str, Any] | None = None

File: Jarvis-main/packages/schemas/contracts.ts
import type { DecisionBand, JobStatus, OutputMode, ZoningDecision } from "./enums";

export interface CategoryComparisonRow {
  category_code: string;
  band: DecisionBand;
  suitability_score: number;
  saturation_score: number;
  opportunity_score: number;
  risk_level: string;
  preferred_floor_type?: string;
  preferred_unit_size?: string;
  reason_summary: string;
  confidence: number;
  suggested_alternative?: string | null;
}

export interface UnsuitableCategoryCard {
  category_code: string;
  rejection_level: "reject" | "high_caution" | "watchlist";
  poor_fit_reason: string;
  saturation_reason?: string;
  demographic_mismatch_reason?: string;
  location_unit_mismatch_reason?: string;
  zoning_caution?: string;
  pricing_burden_warning?: string;
  alternatives: string[];
  operator_explanation: string;
  customer_explanation: string;
  confidence: number;
}

export interface ReliabilityIndicator {
  module_name: string;
  overall_confidence_score: number;
  freshness_score: number;
  source_quality_score: number;
  coverage_score: number;
  model_stability_score: number;
  missing_data_warning: boolean;
  observed_ratio: number;
  estimated_ratio: number;
  manual_review_recommended: boolean;
  explanation_note: string;
}

export interface ValidationAssessment {
  assessment_id: string;
  tenant_validation_score: number;
  decision_band: DecisionBand;
  risk_level: string;
  suitability_score: number;
  saturation_score: number;
  opportunity_score: number;
}

export interface PricingRecommendation {
  unit_id: string;
  recommended_deposit_krw: number;
  recommended_monthly_rent_krw: number;
  lower_bound_krw?: number;
  upper_bound_krw?: number;
  pricing_confidence_score: number;
  assumptions: Record<string, unknown>;
}

export interface ZoningScreeningResult {
  rule_id: string;
  decision_status: ZoningDecision;
  operator_message: string;
  customer_message: string;
  legal_basis_text?: string;
  legal_reference_code?: string;
  review_required: boolean;
  review_owner_user_id?: string | null;
}

export interface CRMLeadSummary {
  lead_id: string;
  lead_type: string;
  lead_stage: string;
  lead_score: number;
  validation_assessment_id?: string | null;
  report_run_id?: string | null;
  next_followup_at?: string | null;
}

export interface ReportSummary {
  report_run_id: string;
  mode: OutputMode;
  status: JobStatus;
  section_status: { section_key: string; ready: boolean }[];
  evidence_coverage_ratio: number;
  html_uri?: string;
  pdf_uri?: string;
}

export interface AnalysisJobStatus {
  job_id: string;
  module: string;
  status: JobStatus;
  percent_progress: number;
  partial_available: boolean;
  partial_payload_uri?: string;
  retry_token?: string;
  error?: { code: string; message: string; retryable: boolean } | null;
}
